text/FilipBartodziej/opis.tex

```diff
@@ -0,0 +1,188 @@
+\documentclass[
+12pt
+]{article}
+\usepackage[T1]{fontenc}
+\usepackage[utf8]{inputenc}
+\usepackage{amsthm}
+\usepackage{amsmath} 
+\usepackage{hyperref}
+\usepackage{cleveref}
+\usepackage{sectsty}
+\usepackage{algorithmicx,algpseudocode}
+
+\sectionfont{\fontsize{15}{15}\selectfont}
+
+\newtheorem{theorem}{Theorem}
+\newtheorem{corollary}{Corollary}[theorem]
+\newtheorem{lemma}[theorem]{Lemma}
+
+\theoremstyle{definition}
+\newtheorem{definition}{Definition}
+
+\theoremstyle{remark}
+\newtheorem*{remark}{Remark}
+
+
+\algnewcommand\algorithmicinput{\textbf{INPUT:}}
+\algnewcommand\Input{\item[\algorithmicinput]}
+
+\algnewcommand\algorithmicoutput{\textbf{OUTPUT:}}
+\algnewcommand\Output{\item[\algorithmicoutput]}
+
+\begin{document}
+
+\begin{center}
+\Huge Algorytm small-large
+\end{center}
+
+
+\section*{Tablica sufiksów}
+Dla słowa s niech $T_i$ oznacza sufiks rozpoczynający się na indeksie i, a $t_i$ to znak na i-tej pozycji. Tablica sufiksów s to tablica sufiksów s (reprezentowanych przez indeksy), to tablica posortowana według porządku leksykograficznego na sufiksach. Tablica ta przydaje się np. przy budowaniu drzew sufiksowych. Poniżej przedstawiony jest algorytm budujący tablicę sufiksów w czasie i pamięci liniowej.
+
+W poniższych rozważaniach bez straty ogólności zakładamy, że każde słowo s kończy się znakiem unikalnym i najmniejszym w słowie. Zakładamy również że alfabet to \{1,2,...,n\}.
+
+\section*{SL - kategoryzacja sufiksów}
+
+Na początek przedstawimy kategoryzacje prefiksów na typy S (small) i L (large). Główna idea algorytmu jest oparta właśnie na tej kategoryzacji i jej własnościach.
+
+\begin{definition}
+Sufiks i jest typu S/L, gdy jest mniejszy/większy od sufiksu i+1. Sufiks n jest S i L (zależnie od potrzeby).
+\end{definition}
+
+Można, używając poniższego faktu, w czasie i pamięci liniowej dokonać kategoryzacji sufiksów na typy S/L.
+
+\begin{remark}
+Dla sufiksu $T_i \ (i<n)$:
+
+\begin{itemize}
+\item jeśli $t_i \neq t_{i+1}$, to wystarczy porównać jeden znak aby dowiedzieć się jakiego typu jest $T_i$
+\item
+jeśli $t_{i} = t_{i+1}$, to znajdź pierwszy $j > i$, taki że $t_j \neq t_i$, wtedy \\
+$ t_j > t_i \Rightarrow T_i,T_{i+1},...,T_{j-1}$ są typu S \\
+$ t_j < t_i \Rightarrow T_i,T_{i+1},...,T_{j-1}$ są typu L
+\end{itemize}
+
+\end{remark}
+
+Przedstawmy również ważną własność kategoryzacji SL.
+
+\begin{lemma}\label{SL_order}
+Sufiks typu S jest leksykograficznie większy od każdego sufiksu typu L, który rozpoczyna się tym samym znakiem.
+\end{lemma}
+\begin{proof}
+Dowód nie wprost - załóżmy że istnieją sufiksy: $T_i = c \alpha c_1 \beta$ typ S i $T_j = c\alpha c_2 \gamma$ typ L, $c_1 \neq c_2$, takie że $T_i < T_j$.
```

Comment:
For clarity please add that $c, c_1, c_2 \in \A$, $\alpha, \beta, \gamma \in \A^*$ - I know it's fairly obvious, but just in case not to confuse any future reader.